### PR TITLE
intial work on #2647 - caching

### PR DESCRIPTION
**Pull request by lampwins**
### Fixes: #2647 

This implements caching via Redis using `django-redis`. This officially makes Redis required for NetBox.

I also implemented basic support for `django-prometheus` because I strongly feel that with the addition of caching, we need to provide more visibility into application performance. This will also aid in development down the road.

This still needs to have docs writen for caching in general and the new configuration settings.

As for review, much of this is rather boring; the only real meat here is in settings.py.

### CODE REVIEW

**Comment by jeremystretch** [comment]
> I also implemented basic support for `django-prometheus`

How is this related to caching? This would need to have its own issue, discussion, and separate PR.

**Comment by jeremystretch** [comment]
> psycopg2.OperationalError: FATAL:  database "netbox" does not exist

That's a new one. Come on, Travis, get it together.

File: base_requirements.txt [hunk 1]
```diff
@@ -18,6 +18,14 @@ django-filter
 # https://github.com/django-mptt/django-mptt
 django-mptt
 
+# Prometheus metrics client implementation for Django
+# https://github.com/korfuri/django-prometheus
+django-prometheus
```
Comment (jeremystretch) [inline base_requirements.txt]: Let's break out the prometheus stuff from the caching work. It really needs to be its own FR with discussion separate from #2647. I'd also much prefer to evaluate caching on its own prior to the introduction of new middleware, which could complicate things.

File: netbox/circuits/urls.py [hunk 2]
```diff
@@ -2,48 +2,49 @@
...
+from utilities.urls import cached
 from . import views
 from .models import Circuit, CircuitTermination, CircuitType, Provider
 
 app_name = 'circuits'
 urlpatterns = [
 
     # Providers
-    url(r'^providers/$', views.ProviderListView.as_view(), name='provider_list'),
+    url(r'^providers/$', cached(views.ProviderListView.as_view()), name='provider_list'),
```
Comment (jeremystretch) [inline netbox/circuits/urls.py]: It's probably cleanest to keep caching functions out of `urls.py`. We can wrap `get()` inside ObjectListView to knock out all the object lists, and wrap `get()` for each of the individual object views.

File: netbox/netbox/configuration.example.py [hunk 3]
```diff
@@ -50,6 +60,18 @@
 # BASE_PATH = 'netbox/'
 BASE_PATH = ''
 
+# Cache timeout in seconds. Set to `None` to enforce an infinate timeout. Set to 0 to dissable caching by immediatly
```
Comment (jeremystretch) [inline netbox/netbox/configuration.example.py]: Do we want to support an infinite timeout? Not sure what the implications would be.

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/netbox/configuration.example.py (hunk 3 above)
Comment (lampwins) [reply netbox/netbox/configuration.example.py]: As discussed in slack, this is just what the Django Caching API supports out of the box. We will leave it be for now.

**Pull request merged**